File: cpp_exercises/credit.cpp
#include <iostream>
#include <cmath>

using namespace std;

bool validateCard(long long cardNumber);
int isValidCard(long long cardNumber);

int main()
{
    long long cardNumber;
    do
    {
        cout << "digite somente os números do seu cartão: ";
        cin >> cardNumber;
    } while (cardNumber <= 0);

    if (validateCard(cardNumber))
    {
        isValidCard(cardNumber);
    }
    else
    {
        cout << "INVÁLIDO";
    }
}

bool validateCard(long long cardNumber)
{

    int sum = 0;
    bool isSecondNumber = false;

    while (cardNumber != 0)
    {
        int digit = cardNumber % 10;

        if (isSecondNumber)
        {
            digit *= 2;
            sum += digit / 10 + digit % 10;
        }
        else
        {
            sum += digit;
        }

        isSecondNumber = !isSecondNumber;
        cardNumber /= 10;
    }

    if (sum % 10 != 0)
    {
        return false;
    }
    else
    {
        return true;
    }
}

int isValidCard(long long cardNumber)
{
    long long size = trunc(log10(cardNumber)) + 1;
    int firstDigit;
    while (cardNumber != 0)
    {
        firstDigit = cardNumber % 10;
        cardNumber = cardNumber / 10;
    }

    if (size == 15 && firstDigit == 3)
    {
        cout << "VÁLIDO\n";
        cout << "AMEX\n";
    }
    else if (size == 16 && firstDigit == 5)
    {
        cout << "VÁLIDO\n";
        cout << "MASTERCARD\n";
    }
    else if (size == 13 || size == 16 && firstDigit == 4)
    {
        cout << "VÁLIDO\n";
        cout << "VISA\n";
    }
    else
    {
        cout << "INVÁLIDO\n";
    }
    return 0;
}

/*examples of valid cards:

Mastercard: 5555555555554444,
            5105105105105100

Visa: 4012888888881881,
      4222222222222

Amex: 378282246310005,
      378734493671000
*/

File: python_exercises/credit.py
from math import trunc
import re

card_number = input("Número do cartão: ")
card_size = len(card_number)

def validate_card(card):
    sum = 0
    card = card[::-1]
    
    for x in range(1, card_size, 2):
        number = int(card[x]) * 2
        if number >= 10:
            sum += trunc(number / 10) + (number % 10)
        else:
            sum += number

    for x in range(0, card_size, 2):
        sum += int(card[x])

    if sum % 10 == 0:
        return True
    else:
        return False

def detect_card_brand():
    amex = re.search('^(34|37)', card_number)
    master = re.search('(51|52|53|54|55)', card_number)
    visa = re.search('^4', card_number)

    if validate_card(card_number) != True:
        print("Cartão invalido")
        return 0
    
    if amex and card_size == 15:
        print("AMEX")
    
    elif master and card_size == 16:
        print("MASTERCARD")

    elif visa and (card_size == 13 or card_size == 16):
        print("VISA")

    print("VÁLIDO")


detect_card_brand()